proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,319 @@
+  associatedtype Element where AsyncIterator.Element == Element
+  associatedtype Failure where AsyncIterator.Failure == Failure
+  __consuming func makeAsyncIterator() -> AsyncIterator
+}
+```
+
+This allows the use of `AsyncSequence` with both opaque types (`some
+AsyncSequence<String, any Error>`) and existential types (`any
+AsyncSequence<Image, NetworkError>`). 
+
+### AsyncThrowingStream
+
+The non-throwing variant `AsyncStream` cannot adopt typed throws
+since it does not have a generic `Failure` parameter; however,
+the throwing variant `AsyncThrowingStream` already contains a `Failure`
+generic parameter and can be extended to adopt typed throws.
+The current initializers of `AsyncThrowingStream` are constraining
+the `Failure` type to `Error`.
+
+```swift
+public init(
+  _ elementType: Element.Type = Element.self,
+  bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded,
+  _ build: (Continuation) -> Void
+) where Failure == Error
+
+public init(
+  unfolding produce: @escaping () async throws -> Element?
+) where Failure == Error 
+```
+
+We propose to add two new initializers that drop the `Failure` constraint:
+
+```swift
+public init(
+  _ elementType: Element.Type = Element.self,
+  _ failureType: Failure.Type = Failure.self,
+  bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded,
+  _ build: (Continuation) -> Void
+)
+
+public init(
+  unfolding produce: @escaping () async throws(Failure) -> Element?
+)
+```
+
+Furthermore, we propose to add a new `makeStream` method that
+similarly drop the `Failure` constraint:
+
+```swift
+@backDeployed(before: SwiftStdlib 5.9)
+public static func makeStream(
+    of elementType: Element.Type = Element.self,
+    throwing failureType: Failure.Type = Failure.self,
+    bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded
+) -> (stream: AsyncThrowingStream<Element, Failure>, continuation: AsyncThrowingStream<Element, Failure>.Continuation) {
+  var continuation: AsyncThrowingStream<Element, Failure>.Continuation!
+  let stream = AsyncThrowingStream<Element, Failure>(bufferingPolicy: limit) { continuation = $0 }
+  return (stream: stream, continuation: continuation!)
+}
+```
+
+### Throwing transformation sequences
+
+- AsyncThrowingCompactMapSequence
+- AsyncThrowingDropWhileSequence
+- AsyncThrowingFilterSequence
+- AsyncThrowingFlatMapSequence
+- AsyncThrowingMapSequence
+- AsyncThrowingPrefixWhileSequence
+
+TODO: All of those don't have a generic error type so we can't adopt typed throws for them.
+
+## Source compatibility
+
+TODO
+
+## Effect on ABI stability
+
+The ABI between an function with an untyped throws and one that uses typed
+throws will be different, so that typed throws can benefit from knowing the
+precise type. For most of the proposed changes, an actual ABI break can
+be avoided because the implementations can make use of
+[`@backDeploy`](https://github.com/apple/swift-evolution/blob/main/proposals/0376-function-back-deployment.md).
```

Comment:
With `withTaskGroup` we're able to introduce new methods by resorting to backwards compatible impls with their old `_silgen_name` names -- so for methods we have a good way forward here.

Not sure if worth mentioning